server/filestore.go

```diff
@@ -4517,43 +4524,71 @@ func (fs *fileStore) enforceMsgPerSubjectLimit(fireCallback bool) {
 		// Rebuild fs state too.
 		fs.rebuildStateLocked(nil)
 		// Need to redo blocks that need attention.
-		needAttention = make(map[string]*psi)
+		needAttention.Empty()
+		fblk, lblk = uint32(math.MaxUint32), uint32(0)
 		fs.psim.IterFast(func(subj []byte, psi *psi) bool {
 			if psi.total > maxMsgsPer {
-				needAttention[string(subj)] = psi
+				needAttention.Insert(subj, psi)
+			}
+			if psi.fblk < fblk {
+				fblk = psi.fblk
+			}
+			if psi.lblk > lblk {
+				lblk = psi.lblk
 			}
 			return true
 		})
 	}
 
+	// If nothing to do then stop.
+	if fblk == math.MaxUint32 || lblk == 0 {
+		return
+	}
+
 	// Collect all the msgBlks we alter.
 	blks := make(map[*msgBlock]struct{})
 
 	// For re-use below.
 	var sm StoreMsg
 
-	// Walk all subjects that need attention here.
-	for subj, info := range needAttention {
-		total, start, stop := info.total, info.fblk, info.lblk
-
-		for i := start; i <= stop; i++ {
-			mb := fs.bim[i]
-			if mb == nil {
-				continue
+	done := make(chan struct{})
+	defer close(done)
+	i := fblk
+	go func() {
+		t := time.NewTicker(10 * time.Second)
+		printed := false
+		for {
+			select {
+			case <-t.C:
+				fs.warn("Enforce message per subject limit %d progress: %d/%d blocks, %d subjects remaining", maxMsgsPer, i-fblk, lblk-fblk+1, needAttention.Size())
+				printed = true
+			case <-done:
+				if printed {
+					fs.warn("Enforce message per subject limit %d complete")
```

Comment:
measure how long it took? `took` etc... so that can filter out in the logs like other similar entries